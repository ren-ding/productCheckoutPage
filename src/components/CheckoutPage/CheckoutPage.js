import React, {Component} from 'react';

export default class CheckoutPage extends Component {
    render() {
        return (
            <div>Checkout page</div>
        );
    }
}