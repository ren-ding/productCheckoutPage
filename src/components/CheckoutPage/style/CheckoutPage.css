.checkout-container {
    display: flex;
}

h1 {
    margin-top: 30px !important;
    text-align: center;
}

.checkout-left-container {
    flex: 3;
    margin: 20px 20px 20px 10px;
}

.checkout-right-container {
    flex: 2;
    margin: 75px 20px 10px 10px;
}