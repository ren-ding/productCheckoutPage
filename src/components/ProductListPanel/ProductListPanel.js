import React, {Component} from 'react';
import PropTypes from 'prop-types';
import ProductSegment from '../ProductSegment';

export default class ProductListPanel extends Component {
    render() {
        return (
            <div className='productlist-wrapper'>
                <h2>ProductListPanel</h2>
                {this.props.products.map(this.renderProductComponent)}
            </div>
        );
    }

    renderProductComponent = (product, index) => (
        <ProductSegment
            key={index}
            product={product}
            addToCheckout={this.props.addToCheckout}
        />
    );
}

ProductListPanel.propTypes = {
    products: PropTypes.array.isRequired,
    addToCheckout: PropTypes.func.isRequired
};