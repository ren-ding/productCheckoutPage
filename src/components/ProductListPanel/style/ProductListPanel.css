.productlist-wrapper {
    margin-top: 15px;
}