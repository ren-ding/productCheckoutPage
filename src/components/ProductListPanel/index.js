import ProductListPanel from './ProductListPanel';
export default ProductListPanel;