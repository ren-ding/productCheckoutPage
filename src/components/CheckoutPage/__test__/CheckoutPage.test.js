import React from 'react';
import CheckoutPage from '../CheckoutPage';
import ProductListPanel from '../../ProductListPanel';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
 
Enzyme.configure({ adapter: new Adapter() });

describe('CheckoutPage', ()=> {
    let checkoutPage;
    let props;

    beforeEach(()=> {
        props = {
            fetchDataForCheckoutPage: () => {},
            submitPromotionCode: () => {}
        }

        checkoutPage = () => shallow(<CheckoutPage {...props} />, { lifecycleExperimental: true });        
    });

    describe('render', () => {
       it('should render to match snapshot', () => {
           const tree = renderer.create(<CheckoutPage {...props}/>).toJSON();
           expect(tree).toMatchSnapshot();
       }); 
    });

    describe('CheckoutPage did load', ()=> {
        const mockProducts = [{
            productId: 'wf',
            productName:'Workflow',
            price:199.99
        }];

        it('should retrieve products', ()=> {
            const mockFetchingData = jest.fn((setProducts)=> {
                setProducts(mockProducts);
            });
            props.fetchDataForCheckoutPage = mockFetchingData;
            const page = checkoutPage();
            expect(mockFetchingData.mock.calls.length).toBe(1);
            expect(page.state('products')).toStrictEqual(mockProducts);
        });
    });

    describe('trigger add a product to checkoutlist', ()=> {
        describe('the will add product does not exist in checkout list', ()=> {
            it('should add this product into the checkout list with quantity set to be 1', ()=> {
                const page = checkoutPage();
                page.instance().addToCheckout('wf');
                expect(page.state('checkoutProducts')).toStrictEqual([ { productId: 'wf', quantity: 1 } ]);
            });
        });

        describe('this product already in the checkout list', ()=> {
            it('should increase its quantity to one more', ()=> {
                const page = checkoutPage();
                page.setState({checkoutProducts: [ { productId: 'wf', quantity: 1 }, { productId: 'docgen', quantity: 1 } ]});
                page.instance().addToCheckout('wf');
                expect(page.state('checkoutProducts')).toStrictEqual([ { productId: 'wf', quantity: 2 },
                                                                       { productId: 'docgen', quantity: 1 }
                                                                     ]);
            });
        });

    });

    describe('trigger remove product from checkoutlist', ()=> {
        it('should remove the product from checkoutlist', ()=> {
            const page = checkoutPage();
            page.setState({checkoutProducts: [ { productId: 'wf', quantity: 2 }, { productId: 'docgen', quantity: 1 } ]});
            page.instance().removeFromCheckout('wf');
            expect(page.state('checkoutProducts')).toStrictEqual([ { productId: 'docgen', quantity: 1 }]);
        });

    });
});