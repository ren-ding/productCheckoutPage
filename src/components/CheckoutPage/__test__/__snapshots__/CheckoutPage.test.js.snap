// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutPage render should render to match snapshot 1`] = `
<div
  className="ui container"
>
  <h1>
    Checkout page
  </h1>
  <div
    className="checkout-container"
  >
    <div
      className="checkout-left-container"
    >
      <div
        className="ui action labeled input"
      >
        <div
          className="ui label"
        >
          Promotion Code:
        </div>
        <input
          type="text"
        />
        <button
          className="ui teal button code-submit"
          onClick={[Function]}
        >
          Submit Code
        </button>
      </div>
    </div>
    <div
      className="checkout-right-container"
    />
  </div>
</div>
`;
